File: go/bob/bob.go
// Package bob contains functions to determine the response given a string
package bob

import (
	"strings"
	"unicode"
)

const (
	defaultResponse         = "Whatever."
	shoutResponse           = "Whoa, chill out!"
	questionResponse        = "Sure."
	shoutedQuestionResponse = "Calm down, I know what I'm doing!"
	emptyStatementResponse  = "Fine. Be that way!"
)

// Hey takes a string and returns a response based on the formatting of the string
func Hey(remark string) string {
	trimmedRemark := strings.TrimSpace(remark)
	response := defaultResponse
	if trimmedRemark == "" {
		response = emptyStatementResponse
	} else if IsQuestion(trimmedRemark) {
		if IsShouting(trimmedRemark) && HasLetters(trimmedRemark) {
			response = shoutedQuestionResponse
		} else {
			response = questionResponse
		}
	} else if IsShouting(trimmedRemark) && HasLetters(trimmedRemark) {
		response = shoutResponse
	}
	return response
}

// IsShouting takes a string and determines if it is all caps
func IsShouting(s string) bool {
	for _, r := range s {
		if unicode.IsLetter(r) && unicode.IsLower(r) {
			return false
		}
	}
	return true
}

// IsQuestion takes a string and determines if the last character is a question mark
func IsQuestion(s string) bool {
	return strings.HasSuffix(s, "?")
}

// HasLetters determines if a string contains any letters
func HasLetters(s string) bool {
	for _, r := range s {
		if unicode.IsLetter(r) {
			return true
		}
	}
	return false
}

File: csharp/bob/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    public static string Response(string statement)
    {
        bool hasLetters = statement.Any(char.IsLetter);
        bool hasLowerCaseLetters = statement.Any(char.IsLower);

        bool isQuestion = statement.TrimEnd().EndsWith('?');
        bool isEmpty = String.IsNullOrWhiteSpace(statement);

        bool isShouting = hasLetters && !hasLowerCaseLetters;

        string response = "Whatever.";

        if (isEmpty) 
        {
            response = "Fine. Be that way!";
        } 
        else if (isQuestion && isShouting)
        {
            response = "Calm down, I know what I'm doing!";
        }
        else if (isQuestion)
        {
            response = "Sure.";
        }
        else if (isShouting)
        {
            response = "Whoa, chill out!";
        }
        return response;
    }
}